rmw_fastrtps_shared_cpp/src/rmw_get_topic_endpoint_info.cpp

```diff
@@ -73,246 +77,38 @@ _validate_params(
-  auto impl = static_cast<CustomParticipantInfo *>(node->data);
-  // The GUID of the participant associated with this node
-  const auto & participant_guid = impl->participant->getGuid();
-  const auto & node_name = node->name;
-  const auto & node_namespace = node->namespace_;
-  ::ParticipantListener * slave_target = impl->listener;
-  auto & topic_cache =
-    is_publisher ? slave_target->writer_topic_cache : slave_target->reader_topic_cache;
-  {
-    std::lock_guard<std::mutex> guard(topic_cache.getMutex());
-    const auto & topic_name_to_data = topic_cache().getTopicNameToTopicData();
-    std::vector<rmw_topic_endpoint_info_t> topic_endpoint_info_vector;
-    for (const auto & topic_name : topic_fqdns) {
-      const auto it = topic_name_to_data.find(topic_name);
-      if (it != topic_name_to_data.end()) {
-        for (const auto & data : it->second) {
-          rmw_topic_endpoint_info_t topic_endpoint_info =
-            rmw_get_zero_initialized_topic_endpoint_info();
-          rmw_ret_t ret = _set_rmw_topic_endpoint_info(
-            &topic_endpoint_info,
-            participant_guid,
-            node_name,
-            node_namespace,
-            data,
-            no_mangle,
-            is_publisher,
-            slave_target,
-            allocator);
-          if (ret != RMW_RET_OK) {
-            // Free topic_endpoint_info
-            _handle_topic_endpoint_info_fini(&topic_endpoint_info, allocator);
-            // Free all the space allocated to the previous topic_endpoint_infos
-            for (auto & tinfo : topic_endpoint_info_vector) {
-              _handle_topic_endpoint_info_fini(&tinfo, allocator);
-            }
-            return ret;
-          }
-          // add rmw_topic_endpoint_info_t to a vector
-          topic_endpoint_info_vector.push_back(topic_endpoint_info);
-        }
-      }
-    }
-
-    // add all the elements from the vector to rmw_topic_endpoint_info_array_t
-    auto count = topic_endpoint_info_vector.size();
-    ret = rmw_topic_endpoint_info_array_init_with_size(participants_info, count, allocator);
-    if (ret != RMW_RET_OK) {
-      rmw_error_string_t error_message = rmw_get_error_string();
-      rmw_reset_error();
-      RMW_SET_ERROR_MSG_WITH_FORMAT_STRING(
-        "rmw_topic_endpoint_info_array_init_with_size failed to allocate memory: %s",
-        error_message.str);
-      // Free all the space allocated to the previous topic_endpoint_infos
-      for (auto & tinfo : topic_endpoint_info_vector) {
-        _handle_topic_endpoint_info_fini(&tinfo, allocator);
-      }
-      return ret;
-    }
-    for (auto i = 0u; i < count; i++) {
-      participants_info->info_array[i] = topic_endpoint_info_vector.at(i);
-    }
-    participants_info->count = count;
+  auto common_context = static_cast<rmw_dds_common::Context *>(node->context->impl->common);
+  rmw_dds_common::GraphCache & graph_cache = common_context->graph_cache;
```

Comment:
@ivanpauno can be made `const`?

Reply:
See https://github.com/ros2/rmw_fastrtps/pull/312/commits/41977050b94d1e1498a35ba3f7c393048f0da8d5.